File: apps/server/src/services/sync.ts
import { eq, and, sql } from "drizzle-orm";
import { db, sqlite } from "../db/client.ts";
import { syncState, users } from "../db/schema.ts";
import {
  getGamesPage,
  getPlayer,
  type RawAoe4WorldGame,
} from "./aoe4world.ts";
import {
  ensureMap,
  isoToUnix,
  normalizeCivSlug,
  normalizeKind,
  normalizeMapSlug,
  normalizeResult,
} from "./normalize.ts";
import { emitSync } from "./syncEvents.ts";
import { log } from "../log.ts";

const SYNC_KEY = "all"; // single bucket; per-leaderboard slicing can be added later

let runningForUser: Map<number, Promise<void>> = new Map();

type SyncOpts = { reqId?: string };

export function isSyncRunning(userId: number): boolean {
  return runningForUser.has(userId);
}

export async function runSync(
  userId: number,
  full = false,
  opts: SyncOpts = {},
): Promise<void> {
  const existing = runningForUser.get(userId);
  if (existing) return existing;
  const p = doRunSync(userId, full, opts).finally(() => runningForUser.delete(userId));
  runningForUser.set(userId, p);
  return p;
}

async function doRunSync(
  userId: number,
  full: boolean,
  opts: SyncOpts,
): Promise<void> {
  const startWallMs = performance.now();

  const profileRow = db()
    .select({ profileId: users.aoe4worldProfileId })
    .from(users)
    .where(eq(users.id, userId))
    .get();
  const profileId = profileRow?.profileId ?? null;
  if (!profileId) {
    throw new Error("aoe4world profile not linked");
  }

  const startedTs = Math.floor(Date.now() / 1000);
  upsertSyncState(userId, { lastPolledAt: startedTs });
  emitSync({
    type: "sync.started",
    user_id: userId,
    profile_id: profileId,
    full,
    ts: startedTs,
  });
  log(opts.reqId, `sync.start user=${userId} profile=${profileId} full=${full}`);

  try {
    const state = readSyncState(userId);
    const lastSeenGameId = state.lastSeenGameId;

    let page = 1;
    let importedThisRun = 0;
    let scannedThisRun = 0;
    let highestSeenGameId = lastSeenGameId ?? 0;

    // Scan newest-first (aoe4world default order is started_at desc). For an
    // incremental sync, stop once we encounter a known game — but always
    // re-ingest the most recent few first so late-arriving fields (final
    // result, rating diff) get refreshed on rows we imported while the game
    // was still ongoing.
    const REFRESH_RECENT = 5;
    pageLoop: while (true) {
      const res = await getGamesPage(profileId, { page, limit: 50 });
      if (!res.games || res.games.length === 0) break;
      const pageStart = performance.now();
      for (const raw of res.games) {
        const alreadySeen =
          !full && lastSeenGameId !== null && raw.game_id <= lastSeenGameId;
        if (alreadySeen && scannedThisRun >= REFRESH_RECENT) {
          break pageLoop;
        }
        const wasNew = ingestGame(userId, profileId, raw);
        if (wasNew) importedThisRun += 1;
        scannedThisRun += 1;
        if (raw.game_id > highestSeenGameId) highestSeenGameId = raw.game_id;
      }
      const pageMs = performance.now() - pageStart;
      emitSync({
        type: "sync.page",
        user_id: userId,
        page,
        games_in_page: res.games.length,
        imported_so_far: importedThisRun,
        scanned_so_far: scannedThisRun,
        total_count: null,
        full,
        ts: Math.floor(Date.now() / 1000),
      });
      log(
        opts.reqId,
        `sync.page user=${userId} page=${page} games=${res.games.length} imported_so_far=${importedThisRun} ingest_ms=${pageMs.toFixed(0)}`,
      );
      if (res.games.length < 50) break;
      page += 1;
      if (page > 200) break; // safety
    }

    upsertSyncState(userId, {
      lastSuccessAt: Math.floor(Date.now() / 1000),
      lastError: null,
      lastSeenGameId: highestSeenGameId || null,
    });
    const durationMs = performance.now() - startWallMs;
    emitSync({
      type: "sync.completed",
      user_id: userId,
      imported: importedThisRun,
      last_seen_game_id: highestSeenGameId || null,
      duration_ms: Math.round(durationMs),
      ts: Math.floor(Date.now() / 1000),
    });
    log(
      opts.reqId,
      `sync.done user=${userId} imported=${importedThisRun} pages=${page} duration_ms=${durationMs.toFixed(0)}`,
    );
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    upsertSyncState(userId, { lastError: msg.slice(0, 1000) });
    emitSync({
      type: "sync.error",
      user_id: userId,
      message: msg,
      ts: Math.floor(Date.now() / 1000),
    });
    log(opts.reqId, `sync.error user=${userId} msg=${msg}`);
    throw e;
  }
}

function readSyncState(userId: number) {
  const row = db()
    .select()
    .from(syncState)
    .where(and(eq(syncState.userId, userId), eq(syncState.leaderboard, SYNC_KEY)))
    .get();
  return {
    lastSeenGameId: row?.lastSeenGameId ?? null,
    lastPolledAt: row?.lastPolledAt ?? null,
    lastSuccessAt: row?.lastSuccessAt ?? null,
    lastError: row?.lastError ?? null,
  };
}

function upsertSyncState(
  userId: number,
  patch: {
    lastSeenGameId?: number | null;
    lastPolledAt?: number | null;
    lastSuccessAt?: number | null;
    lastError?: string | null;
  },
): void {
  // SQLite upsert via raw sql for nullable handling
  const existing = sqlite()
    .prepare(
      "SELECT 1 FROM sync_state WHERE user_id = ? AND leaderboard = ?",
    )
    .get(userId, SYNC_KEY);
  if (!existing) {
    sqlite()
      .prepare(
        "INSERT INTO sync_state (user_id, leaderboard, last_seen_game_id, last_polled_at, last_success_at, last_error) VALUES (?, ?, ?, ?, ?, ?)",
      )
      .run(
        userId,
        SYNC_KEY,
        patch.lastSeenGameId ?? null,
        patch.lastPolledAt ?? null,
        patch.lastSuccessAt ?? null,
        patch.lastError ?? null,
      );
    return;
  }
  const fields: string[] = [];
  const values: unknown[] = [];
  if (Object.prototype.hasOwnProperty.call(patch, "lastSeenGameId")) {
    fields.push("last_seen_game_id = ?");
    values.push(patch.lastSeenGameId ?? null);
  }
  if (Object.prototype.hasOwnProperty.call(patch, "lastPolledAt")) {
    fields.push("last_polled_at = ?");
    values.push(patch.lastPolledAt ?? null);
  }
  if (Object.prototype.hasOwnProperty.call(patch, "lastSuccessAt")) {
    fields.push("last_success_at = ?");
    values.push(patch.lastSuccessAt ?? null);
  }
  if (Object.prototype.hasOwnProperty.call(patch, "lastError")) {
    fields.push("last_error = ?");
    values.push(patch.lastError ?? null);
  }
  if (fields.length === 0) return;
  values.push(userId, SYNC_KEY);
  sqlite()
    .prepare(
      `UPDATE sync_state SET ${fields.join(", ")} WHERE user_id = ? AND leaderboard = ?`,
    )
    .run(...values);
}

export function listSyncState(userId: number) {
  return db().select().from(syncState).where(eq(syncState.userId, userId)).all();
}

function ingestGame(
  userId: number,
  selfProfileId: number,
  raw: RawAoe4WorldGame,
): boolean {
  let isNew = false;
  sqlite().transaction(() => {
    // Find self entry
    const allPlayers: Array<{
      teamIdx: number;
      p: RawAoe4WorldGame["teams"][number][number]["player"];
    }> = [];
    raw.teams.forEach((team, teamIdx) => {
      for (const entry of team) {
        allPlayers.push({ teamIdx, p: entry.player });
      }
    });
    const selfEntry = allPlayers.find((x) => x.p.profile_id === selfProfileId);
    if (!selfEntry) return; // skip games we can't find self in

    const mapSlug = normalizeMapSlug(raw.map ?? null);
    ensureMap(mapSlug, raw.map ?? null);

    const myCivSlug = normalizeCivSlug(selfEntry.p.civilization);
    const myResult = normalizeResult(selfEntry.p.result);
    const kind = normalizeKind(raw.kind ?? null);
    const startedAt = isoToUnix(raw.started_at);

    // Upsert game by (user_id, aoe4world_game_id)
    const existing = sqlite()
      .prepare(
        "SELECT id FROM games WHERE user_id = ? AND aoe4world_game_id = ?",
      )
      .get(userId, raw.game_id) as { id: number } | undefined;

    let gameId: number;
    if (existing) {
      gameId = existing.id;
      sqlite()
        .prepare(
          `UPDATE games SET
             duration_seconds = ?,
             map_slug = ?,
             kind = ?,
             leaderboard = ?,
             patch = ?,
             server = ?,
             my_team = ?,
             my_civ_slug = ?,
             my_civ_randomized = ?,
             my_result = ?,
             my_rating = ?,
             my_rating_diff = ?,
             my_mmr = ?,
             raw_payload_json = ?,
             updated_at = unixepoch()
           WHERE id = ?`,
        )
        .run(
          raw.duration ?? null,
          mapSlug,
          kind,
          raw.leaderboard ?? null,
          raw.patch ?? null,
          raw.server ?? null,
          selfEntry.teamIdx,
          myCivSlug,
          selfEntry.p.civilization_randomized === true
            ? 1
            : selfEntry.p.civilization_randomized === false
              ? 0
              : null,
          myResult,
          selfEntry.p.rating ?? null,
          selfEntry.p.rating_diff ?? null,
          selfEntry.p.mmr ?? null,
          JSON.stringify(raw),
          gameId,
        );
    } else {
      const result = sqlite()
        .prepare(
          `INSERT INTO games (
             user_id, source, aoe4world_game_id, started_at, duration_seconds,
             map_slug, kind, leaderboard, patch, server, my_team,
             my_civ_slug, my_civ_randomized, my_result, my_rating, my_rating_diff, my_mmr,
             raw_payload_json, imported_at, created_at, updated_at
           ) VALUES (?, 'aoe4world', ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, unixepoch(), unixepoch(), unixepoch())`,
        )
        .run(
          userId,
          raw.game_id,
          startedAt,
          raw.duration ?? null,
          mapSlug,
          kind,
          raw.leaderboard ?? null,
          raw.patch ?? null,
          raw.server ?? null,
          selfEntry.teamIdx,
          myCivSlug,
          selfEntry.p.civilization_randomized === true
            ? 1
            : selfEntry.p.civilization_randomized === false
              ? 0
              : null,
          myResult,
          selfEntry.p.rating ?? null,
          selfEntry.p.rating_diff ?? null,
          selfEntry.p.mmr ?? null,
          JSON.stringify(raw),
        );
      gameId = Number(result.lastInsertRowid);
      isNew = true;
    }

    // Replace participants
    sqlite().prepare("DELETE FROM game_participants WHERE game_id = ?").run(gameId);
    const insertP = sqlite().prepare(
      `INSERT INTO game_participants (
         game_id, team, is_self, profile_id, name, civ_slug, civ_randomized,
         result, rating, rating_diff, mmr
       ) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
    );
    for (const { teamIdx, p } of allPlayers) {
      insertP.run(
        gameId,
        teamIdx,
        p.profile_id === selfProfileId ? 1 : 0,
        p.profile_id ?? null,
        p.name,
        normalizeCivSlug(p.civilization),
        p.civilization_randomized === true
          ? 1
          : p.civilization_randomized === false
            ? 0
            : null,
        normalizeResult(p.result),
        p.rating ?? null,
        p.rating_diff ?? null,
        p.mmr ?? null,
      );
    }
  })();

  return isNew;
}

export type UserDataCounts = {
  games: number;
  game_notes: number;
  sync_state_rows: number;
};

export type UserAllDataCounts = UserDataCounts & {
  civ_notes: number;
  matchup_notes: number;
  map_notes: number;
};

function countOne(sqlText: string, userId: number): number {
  return (sqlite().prepare(sqlText).get(userId) as { c: number }).c;
}

export function countUserGameData(userId: number): UserDataCounts {
  return {
    games: countOne("SELECT COUNT(*) AS c FROM games WHERE user_id = ?", userId),
    game_notes: countOne(
      "SELECT COUNT(*) AS c FROM game_notes WHERE user_id = ?",
      userId,
    ),
    sync_state_rows: countOne(
      "SELECT COUNT(*) AS c FROM sync_state WHERE user_id = ?",
      userId,
    ),
  };
}

export function countAllUserData(userId: number): UserAllDataCounts {
  return {
    ...countUserGameData(userId),
    civ_notes: countOne(
      "SELECT COUNT(*) AS c FROM civ_notes WHERE user_id = ?",
      userId,
    ),
    matchup_notes: countOne(
      "SELECT COUNT(*) AS c FROM matchup_notes WHERE user_id = ?",
      userId,
    ),
    map_notes: countOne(
      "SELECT COUNT(*) AS c FROM map_notes WHERE user_id = ?",
      userId,
    ),
  };
}

export function wipeUserGameData(userId: number): UserDataCounts {
  const before = countUserGameData(userId);
  sqlite().transaction(() => {
    // game_participants and game_notes cascade via FK ON DELETE CASCADE.
    sqlite().prepare("DELETE FROM games WHERE user_id = ?").run(userId);
    sqlite().prepare("DELETE FROM sync_state WHERE user_id = ?").run(userId);
  })();
  return before;
}

export function wipeAllUserData(userId: number): UserAllDataCounts {
  const before = countAllUserData(userId);
  sqlite().transaction(() => {
    sqlite().prepare("DELETE FROM games WHERE user_id = ?").run(userId);
    sqlite().prepare("DELETE FROM sync_state WHERE user_id = ?").run(userId);
    sqlite().prepare("DELETE FROM civ_notes WHERE user_id = ?").run(userId);
    sqlite().prepare("DELETE FROM matchup_notes WHERE user_id = ?").run(userId);
    sqlite().prepare("DELETE FROM map_notes WHERE user_id = ?").run(userId);
  })();
  return before;
}

export async function linkProfileAndBackfill(
  userId: number,
  profileId: number,
  opts: SyncOpts = {},
): Promise<void> {
  const player = await getPlayer(profileId);
  log(opts.reqId, `link.player_fetched profile_id=${profileId} name=${player.name}`);

  // If switching to a different profile, wipe all per-user game data first.
  // Same-profile re-link or null→profile link is non-destructive.
  const existing = db()
    .select({ profileId: users.aoe4worldProfileId })
    .from(users)
    .where(eq(users.id, userId))
    .get();
  const previousProfileId = existing?.profileId ?? null;
  if (previousProfileId !== null && previousProfileId !== profileId) {
    const wiped = wipeUserGameData(userId);
    log(
      opts.reqId,
      `link.wiped previous_profile=${previousProfileId} new_profile=${profileId} games=${wiped.games} game_notes=${wiped.game_notes} sync_rows=${wiped.sync_state_rows}`,
    );
  }

  db()
    .update(users)
    .set({
      aoe4worldProfileId: profileId,
      displayName: player.name,
      updatedAt: sql`(unixepoch())`,
    })
    .where(eq(users.id, userId))
    .run();
  emitSync({
    type: "link.player_fetched",
    user_id: userId,
    profile_id: profileId,
    display_name: player.name,
    ts: Math.floor(Date.now() / 1000),
  });
  await runSync(userId, true, opts);
}

File: src-tauri/src/services/sync.rs
use crate::db::Db;
use crate::error::{AppError, AppResult};
use crate::events::{emit_sync, unix_secs_now, SyncEvent};
use crate::services::aoe4world::{
    get_games_page, get_player, GamesQueryArgs, RawAoe4WorldGame,
};
use crate::services::normalize::{
    ensure_map, iso_to_unix, normalize_civ_slug, normalize_kind, normalize_map_slug,
    normalize_result,
};
use rusqlite::{params, OptionalExtension};
use std::collections::HashSet;
use std::sync::Mutex;
use std::time::Instant;

const SYNC_KEY: &str = "all";

static IN_FLIGHT_USERS: Mutex<Option<HashSet<i64>>> = Mutex::new(None);

fn try_acquire(user_id: i64) -> bool {
    let mut g = IN_FLIGHT_USERS.lock().expect("sync set poisoned");
    let set = g.get_or_insert_with(HashSet::new);
    set.insert(user_id)
}

fn release(user_id: i64) {
    let mut g = IN_FLIGHT_USERS.lock().expect("sync set poisoned");
    if let Some(set) = g.as_mut() {
        set.remove(&user_id);
    }
}

pub fn is_sync_running(user_id: i64) -> bool {
    let g = IN_FLIGHT_USERS.lock().expect("sync set poisoned");
    g.as_ref().is_some_and(|s| s.contains(&user_id))
}

struct SyncStateRow {
    last_seen_game_id: Option<i64>,
}

fn read_sync_state(db: &Db, user_id: i64) -> AppResult<SyncStateRow> {
    let conn = db.0.lock().expect("db poisoned");
    let row: Option<Option<i64>> = conn
        .query_row(
            "SELECT last_seen_game_id FROM sync_state WHERE user_id = ?1 AND leaderboard = ?2",
            params![user_id, SYNC_KEY],
            |r| r.get::<_, Option<i64>>(0),
        )
        .optional()?;
    Ok(SyncStateRow {
        last_seen_game_id: row.flatten(),
    })
}

#[derive(Default)]
struct StatePatch {
    last_seen_game_id: Option<Option<i64>>,
    last_polled_at: Option<Option<i64>>,
    last_success_at: Option<Option<i64>>,
    last_error: Option<Option<String>>,
}

fn upsert_sync_state(db: &Db, user_id: i64, patch: StatePatch) -> AppResult<()> {
    let conn = db.0.lock().expect("db poisoned");
    let exists: Option<i64> = conn
        .query_row(
            "SELECT 1 FROM sync_state WHERE user_id = ?1 AND leaderboard = ?2",
            params![user_id, SYNC_KEY],
            |r| r.get(0),
        )
        .optional()?;
    if exists.is_none() {
        conn.execute(
            "INSERT INTO sync_state (user_id, leaderboard, last_seen_game_id, last_polled_at, last_success_at, last_error)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
            params![
                user_id,
                SYNC_KEY,
                patch.last_seen_game_id.unwrap_or(None),
                patch.last_polled_at.unwrap_or(None),
                patch.last_success_at.unwrap_or(None),
                patch.last_error.unwrap_or(None),
            ],
        )?;
        return Ok(());
    }
    let mut sets: Vec<&str> = Vec::new();
    let mut vals: Vec<Box<dyn rusqlite::ToSql>> = Vec::new();
    if let Some(v) = patch.last_seen_game_id {
        sets.push("last_seen_game_id = ?");
        vals.push(Box::new(v));
    }
    if let Some(v) = patch.last_polled_at {
        sets.push("last_polled_at = ?");
        vals.push(Box::new(v));
    }
    if let Some(v) = patch.last_success_at {
        sets.push("last_success_at = ?");
        vals.push(Box::new(v));
    }
    if let Some(v) = patch.last_error {
        sets.push("last_error = ?");
        vals.push(Box::new(v));
    }
    if sets.is_empty() {
        return Ok(());
    }
    vals.push(Box::new(user_id));
    vals.push(Box::new(SYNC_KEY));
    let sql = format!(
        "UPDATE sync_state SET {} WHERE user_id = ? AND leaderboard = ?",
        sets.join(", ")
    );
    conn.execute(
        &sql,
        rusqlite::params_from_iter(vals.iter().map(|b| b.as_ref())),
    )?;
    Ok(())
}

pub fn run_sync(app: &tauri::AppHandle, db: &Db, user_id: i64, full: bool) -> AppResult<()> {
    if !try_acquire(user_id) {
        return Ok(());
    }
    let result = do_run_sync(app, db, user_id, full);
    release(user_id);
    result
}

fn fetch_profile_id(db: &Db, user_id: i64) -> AppResult<Option<i64>> {
    let conn = db.0.lock().expect("db poisoned");
    let row: Option<Option<i64>> = conn
        .query_row(
            "SELECT aoe4world_profile_id FROM users WHERE id = ?1",
            params![user_id],
            |r| r.get::<_, Option<i64>>(0),
        )
        .optional()?;
    Ok(row.flatten())
}

fn do_run_sync(app: &tauri::AppHandle, db: &Db, user_id: i64, full: bool) -> AppResult<()> {
    let start = Instant::now();
    let profile_id = fetch_profile_id(db, user_id)?
        .ok_or_else(|| AppError::msg("aoe4world profile not linked"))?;
    let started_ts = unix_secs_now();
    upsert_sync_state(
        db,
        user_id,
        StatePatch {
            last_polled_at: Some(Some(started_ts)),
            ..Default::default()
        },
    )?;
    emit_sync(
        app,
        SyncEvent::SyncStarted {
            user_id,
            profile_id,
            full,
            ts: started_ts,
        },
    );
    log::info!("sync.start user={user_id} profile={profile_id} full={full}");

    let state = read_sync_state(db, user_id)?;
    let last_seen = state.last_seen_game_id;
    let refresh_recent: usize = 5;
    let mut page: i64 = 1;
    let mut imported: usize = 0;
    let mut scanned: usize = 0;
    let mut highest = last_seen.unwrap_or(0);

    let result: AppResult<()> = (|| {
        loop {
            let res = get_games_page(
                profile_id,
                &GamesQueryArgs {
                    page: Some(page),
                    limit: Some(50),
                    ..Default::default()
                },
            )?;
            if res.games.is_empty() {
                break;
            }
            let n = res.games.len();
            let mut stop = false;
            for raw in &res.games {
                let already_seen =
                    !full && last_seen.is_some_and(|ls| raw.game_id <= ls);
                if already_seen && scanned >= refresh_recent {
                    stop = true;
                    break;
                }
                let was_new = ingest_game(db, user_id, profile_id, raw)?;
                if was_new {
                    imported += 1;
                }
                scanned += 1;
                if raw.game_id > highest {
                    highest = raw.game_id;
                }
            }
            emit_sync(
                app,
                SyncEvent::SyncPage {
                    user_id,
                    page: page as u32,
                    games_in_page: n,
                    imported_so_far: imported,
                    scanned_so_far: scanned,
                    total_count: None,
                    full,
                    ts: unix_secs_now(),
                },
            );
            log::info!(
                "sync.page user={user_id} page={page} games={n} imported_so_far={imported}"
            );
            if stop || n < 50 {
                break;
            }
            page += 1;
            if page > 200 {
                break;
            }
        }
        Ok(())
    })();

    match result {
        Ok(()) => {
            upsert_sync_state(
                db,
                user_id,
                StatePatch {
                    last_success_at: Some(Some(unix_secs_now())),
                    last_error: Some(None),
                    last_seen_game_id: Some(if highest > 0 { Some(highest) } else { None }),
                    ..Default::default()
                },
            )?;
            let dur_ms = start.elapsed().as_millis() as u64;
            emit_sync(
                app,
                SyncEvent::SyncCompleted {
                    user_id,
                    imported,
                    last_seen_game_id: if highest > 0 { Some(highest) } else { None },
                    duration_ms: dur_ms,
                    ts: unix_secs_now(),
                },
            );
            log::info!(
                "sync.done user={user_id} imported={imported} pages={page} duration_ms={dur_ms}"
            );
            Ok(())
        }
        Err(e) => {
            let msg = e.to_string();
            let truncated: String = msg.chars().take(1000).collect();
            let _ = upsert_sync_state(
                db,
                user_id,
                StatePatch {
                    last_error: Some(Some(truncated)),
                    ..Default::default()
                },
            );
            emit_sync(
                app,
                SyncEvent::SyncError {
                    user_id,
                    message: msg.clone(),
                    ts: unix_secs_now(),
                },
            );
            log::warn!("sync.error user={user_id} msg={msg}");
            Err(e)
        }
    }
}

fn bool_to_int(b: Option<bool>) -> Option<i64> {
    b.map(|x| if x { 1 } else { 0 })
}

fn ingest_game(
    db: &Db,
    user_id: i64,
    self_profile_id: i64,
    raw: &RawAoe4WorldGame,
) -> AppResult<bool> {
    let mut conn = db.0.lock().expect("db poisoned");
    let tx = conn.transaction()?;
    let mut is_new = false;

    // Collect (team_idx, player) for all players
    let mut all_players: Vec<(i64, &crate::services::aoe4world::RawPlayerRef)> = Vec::new();
    for (idx, team) in raw.teams.iter().enumerate() {
        for entry in team {
            all_players.push((idx as i64, &entry.player));
        }
    }
    let self_entry = all_players
        .iter()
        .find(|(_, p)| p.profile_id == self_profile_id)
        .copied();
    let Some((self_team, self_p)) = self_entry else {
        tx.commit()?;
        return Ok(false);
    };

    let map_slug = normalize_map_slug(raw.map.as_deref());
    ensure_map(&tx, map_slug.as_deref(), raw.map.as_deref())?;

    let my_civ_slug = normalize_civ_slug(&tx, self_p.civilization.as_deref())?;
    let my_result = normalize_result(self_p.result.as_deref());
    let kind = normalize_kind(raw.kind.as_deref());
    let started_at = iso_to_unix(Some(&raw.started_at));
    let raw_json = serde_json::to_string(raw)?;

    let existing_id: Option<i64> = tx
        .query_row(
            "SELECT id FROM games WHERE user_id = ?1 AND aoe4world_game_id = ?2",
            params![user_id, raw.game_id],
            |r| r.get(0),
        )
        .optional()?;

    let game_id: i64 = if let Some(id) = existing_id {
        tx.execute(
            "UPDATE games SET
                duration_seconds = ?1,
                map_slug = ?2,
                kind = ?3,
                leaderboard = ?4,
                patch = ?5,
                server = ?6,
                my_team = ?7,
                my_civ_slug = ?8,
                my_civ_randomized = ?9,
                my_result = ?10,
                my_rating = ?11,
                my_rating_diff = ?12,
                my_mmr = ?13,
                raw_payload_json = ?14,
                updated_at = unixepoch()
             WHERE id = ?15",
            params![
                raw.duration,
                map_slug,
                kind,
                raw.leaderboard,
                raw.patch,
                raw.server,
                self_team,
                my_civ_slug,
                bool_to_int(self_p.civilization_randomized),
                my_result,
                self_p.rating,
                self_p.rating_diff,
                self_p.mmr,
                raw_json,
                id,
            ],
        )?;
        id
    } else {
        tx.execute(
            "INSERT INTO games (
                user_id, source, aoe4world_game_id, started_at, duration_seconds,
                map_slug, kind, leaderboard, patch, server, my_team,
                my_civ_slug, my_civ_randomized, my_result, my_rating, my_rating_diff, my_mmr,
                raw_payload_json, imported_at, created_at, updated_at
             ) VALUES (?1, 'aoe4world', ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12, ?13, ?14, ?15, ?16, ?17, unixepoch(), unixepoch(), unixepoch())",
            params![
                user_id,
                raw.game_id,
                started_at,
                raw.duration,
                map_slug,
                kind,
                raw.leaderboard,
                raw.patch,
                raw.server,
                self_team,
                my_civ_slug,
                bool_to_int(self_p.civilization_randomized),
                my_result,
                self_p.rating,
                self_p.rating_diff,
                self_p.mmr,
                raw_json,
            ],
        )?;
        is_new = true;
        tx.last_insert_rowid()
    };

    tx.execute("DELETE FROM game_participants WHERE game_id = ?1", params![game_id])?;
    {
        let mut stmt = tx.prepare(
            "INSERT INTO game_participants (
                game_id, team, is_self, profile_id, name, civ_slug, civ_randomized,
                result, rating, rating_diff, mmr
             ) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11)",
        )?;
        for (team_idx, p) in &all_players {
            let civ_slug = normalize_civ_slug(&tx, p.civilization.as_deref())?;
            stmt.execute(params![
                game_id,
                team_idx,
                if p.profile_id == self_profile_id { 1i64 } else { 0i64 },
                p.profile_id,
                p.name,
                civ_slug,
                bool_to_int(p.civilization_randomized),
                normalize_result(p.result.as_deref()),
                p.rating,
                p.rating_diff,
                p.mmr,
            ])?;
        }
    }

    tx.commit()?;
    Ok(is_new)
}

pub fn link_profile_and_backfill(
    app: &tauri::AppHandle,
    db: &Db,
    user_id: i64,
    profile_id: i64,
) -> AppResult<()> {
    let player = get_player(profile_id)?;
    log::info!("link.player_fetched profile_id={profile_id} name={}", player.name);

    let previous = fetch_profile_id(db, user_id)?;
    if let Some(prev) = previous {
        if prev != profile_id {
            wipe_user_game_data(db, user_id)?;
            log::info!("link.wiped previous_profile={prev} new_profile={profile_id}");
        }
    }

    {
        let conn = db.0.lock().expect("db poisoned");
        conn.execute(
            "UPDATE users SET aoe4world_profile_id = ?1, display_name = ?2, updated_at = unixepoch() WHERE id = ?3",
            params![profile_id, player.name, user_id],
        )?;
    }

    emit_sync(
        app,
        SyncEvent::LinkPlayerFetched {
            user_id,
            profile_id,
            display_name: player.name,
            ts: unix_secs_now(),
        },
    );

    run_sync(app, db, user_id, true)
}

pub fn wipe_user_game_data(db: &Db, user_id: i64) -> AppResult<()> {
    let mut conn = db.0.lock().expect("db poisoned");
    let tx = conn.transaction()?;
    tx.execute("DELETE FROM games WHERE user_id = ?1", params![user_id])?;
    tx.execute("DELETE FROM sync_state WHERE user_id = ?1", params![user_id])?;
    tx.commit()?;
    Ok(())
}
